### ISSUE TITLE
ViM ignores Escape keys on tmux if I'm too fast

### ISSUE THREAD

**Issue by cecton** [tmux/tmux#907]
Hello,

I'm using Arch Linux and xterm, I recently upgraded tmux from 2.3 to 2.4 and encountered multiple issues at the point I had to revert back to tmux 2.3. I reproduced one of the blocking issue using tmux's master branch:

It's hard to reproduce and is related to ViM. If I'm "too fast", some commands are ignored.

Here is how to reproduce:

 - Go to insert mode (o)
 - Add random text
 - Press Escape to exit insert mode
 - Press Up to move to the previous line

On tmux 2.3, you will be on normal mode on the previous line. On tmux 2.4, the Escape has been ignored, you will be in insert mode on the current line. I double checked on master, I can reproduce it easily.

This particular issue makes tmux 2.4 unusable for me because I'm using vim heavily.

Server logs: https://pastebin.com/NxJPTCkC

**Comment by nicm** [comment]
Are you sure you pressed Up in this log?

**Comment by cecton** [comment]
Double checked: https://pastebin.com/XNzYxM9J

Yes I'm sure. I pressed Escape (ignored) then Up (ignored) then waited, then Up (it works) then Escape (it works).

**Comment by nicm** [comment]
You are pressing Up too quickly, before the escape timer expires, so the key is M-Up and not Up. Now that xterm keys is on, this is a different escape sequence which presumably vim is ignoring.

Someone else reported a problem with this and vi because it doesn't understand xterm-style keys at all (whereas I think vim understands them but ignores them). You can try this change: [x.diff.txt](https://github.com/tmux/tmux/files/972993/x.diff.txt)

**Comment by cecton** [comment]
Seems to work perfectly: https://pastebin.com/dJqx8Fsg

**Comment by bug** [comment]
@cecton thanks, I didn't even suspect tmux to be the culprit of that ugly behaviour in vim. I'm facing the exact same issue and reverting to tmux 2.3 fixes the problem.

**Comment by markbook2** [comment]
It may be not related but the following setting from tmux.conf can help
```
set-option -s escape-time 0
```
Then kill tmux server (`tmux kill-server`) and run it again to reload config.

**Comment by cecton** [comment]
@markbook2 I just tested your solution and I confirm it works too.

@nicm what do you think is the good solution?

**Comment by nicm** [comment]
You can set escape-time to 0 if you like, if you never use tmux remotely it will probably never be needed. Or make it lower if your network is fast.

**Comment by nicm** [comment]
Setting it to anything nonzero will probably not help you actually because if it is nonzero tmux will still recognise Escape+Left as M-Left. If you set it to 0 it will not.

**Comment by cecton** [comment]
@nicm I don't think I understand the difference tbh 😞 I'm running tmux on remote SSH sessions, rarely from inside tmux. What I mean is: what is the right solution? Are you going to merge the diff you provided me to master and it will be fixed for the next release or the code is better as it is now and it's best that vi/vim users use the set escape-time 0?

**Comment by nicm** [comment]
If you run tmux remotely with escape-time 0 there is a risk that keys like M-x will be split into two packets (\033 and x) so tmux will see them as two keys. It only matters if you bind any meta keys or functions keys (F1, Left, etc) in tmux itself.

That's why I recommend you use escape-time with a small value like 10 or 20 or 50 or something, which should be enough on a fast network. But it won't help your problem then, because tmux will still see Escape then Up as M-Up, but with xterm-keys now on, M-Up is sent as \033[1;3A instead of \033\033[A, so vim sees it as M-Up not as Escape then Up.

Best for you may actually just be to turn xterm-keys off, but eventually we will remove xterm-keys so that will only help for the next few releases.

I will apply the diff but not until next week because I am waiting to see if someone else gets back to me about it.

**Comment by hcraT** [comment]
I think I have encountered another instance of the same problem.

I use readline vi-mode system-wide. When I use a REPL (for example with R or python) and when in insert mode I press Escape to get back to normal mode sometimes part of the line gets deleted and readline remains in insert mode.

I've downloaded the last source code and applied your patch and the problem seems to be fixed.

**Comment by nicm** [comment]
I have applied this to OpenBSD now, it will be in GitHub soonish. Thanks.

Any further problems please open a new issue.

**Comment by tombh** [comment]
I may have found a problem with this fix. Sorry not to open a new issue, but I haven't dug in enough and just wanted some feedback first. I'm using the vim-tmux-navigator plugin and binding ALT-ARROW keys for all pane navigations. From what I can tell, in the switch from tmux 2.4 to 2.5, Vim no longer correctly receives `send-keys M-[arrow]`. I'm imaging that something in the fix here changed the way `send-keys` sends ALT+ARROW?

**Comment by nickdiego** [comment]
I was experiencing the same issue mentioned by @tombh (on Arch Linux, with tmux 2.5 + vim with vim-tmux-navigator plugin), just downgraded to tmux 2.4 and the issue doesn't occur anymore.

**Comment by tombh** [comment]
@nickdiego @jakesjews I just tried master and the problem is fixed for me. It may have been https://github.com/tmux/tmux/commit/a00b0d13ede4a0f3743606c79a38eada403f77f9 but I haven't bisected so can't be sure.
